### PR TITLE
Filesystem/Win32: Recreate hardlinks on fs::pending_file

### PR DESCRIPTION
**Pull request by elad335**
Attempting to address #16255

### CODE REVIEW

File: Utilities/File.cpp [hunk 1]
```diff
@@ -2486,16 +2485,62 @@ bool fs::pending_file::commit(bool overwrite)
 		disp.DeleteFileW = false;
 		ensure(SetFileInformationByHandle(file.get_handle(), FileDispositionInfo, &disp, sizeof(disp)));
 	}
+
+	std::vector<std::wstring> hardlinkPaths;
```
Comment (Megamouse) [inline Utilities/File.cpp:2489]: Scnakecase

File: Utilities/File.cpp [hunk 2]
```diff
@@ -2486,16 +2485,62 @@ bool fs::pending_file::commit(bool overwrite)
...
+
+	const auto ws1 = to_wchar(m_path);
+
+	const HANDLE fileHandle = CreateFileW(ws1.get(), GENERIC_READ, FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE, nullptr, OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, nullptr);
```
Comment (Megamouse) [inline Utilities/File.cpp:2493]: Snakecase

File: Utilities/File.cpp [hunk 3]
```diff
@@ -2486,16 +2485,62 @@ bool fs::pending_file::commit(bool overwrite)
...
+
+	while (fileHandle != INVALID_HANDLE_VALUE)
+	{
+		// Get file ID (used to check for hardlinks)
+		BY_HANDLE_FILE_INFORMATION fileInfo;
```
Comment (Megamouse) [inline Utilities/File.cpp:2498]: Snakecase

File: Utilities/File.cpp [hunk 4]
```diff
@@ -2486,16 +2485,62 @@ bool fs::pending_file::commit(bool overwrite)
...
+
+		if (!GetFileInformationByHandle(fileHandle, &fileInfo) || fileInfo.nNumberOfLinks == 1)
+		{	
+			CloseHandle(fileHandle);
+			break;
+		}
+
+		// Buffer for holding link name
+		WCHAR linkNameBuffer[MAX_PATH];
+		HANDLE findHandle = FindFirstFileNameW(ws1.get(), 0, nullptr, linkNameBuffer);
```
Comment (RipleyTom) [inline Utilities/File.cpp:2508]: Two things:
-Limiting to MAX_PATH on windows has been a problem in the past if I recall, we should avoid that if possible
-Is nullptr for StringLength even valid, documentation doesn't mention it

Comment (elad335) [reply Utilities/File.cpp]: Fixed

**Comment by digant73** [comment]
maybe the fix for #16245 (reported in that PR) could be also provided in this PR

File: Utilities/File.cpp [hunk 5]
```diff
@@ -2486,16 +2485,87 @@ bool fs::pending_file::commit(bool overwrite)
...
+
+		// Buffer for holding link name
+		std::wstring link_name_buffer(MAX_PATH);
+		LDWORD buffer_size{};
+		HANDLE find_handle = INVALID_HANDLE_VALUE;
+
+		while (true)
+		{
+			buffer_size = link_name_buffer.size() - 1;
+			findHandle = FindFirstFileNameW(ws1.get(), 0, &buffer_size, link_name_buffer.data());
```
Comment (Megamouse) [inline Utilities/File.cpp:2514]:
```suggestion
			find_handle = FindFirstFileNameW(ws1.get(), 0, &buffer_size, link_name_buffer.data());
```

**Pull request merged**